Sherlock/replicator.h

```diff
@@ -0,0 +1,296 @@
+/*******************************************************************************
+The MIT License (MIT)
+
+Copyright (c) 2016 Grigory Nikolaenko <nikolaenko.grigory@gmail.com>
+
+Permission is hereby granted, free of charge, to any person obtaining a copy
+of this software and associated documentation files (the "Software"), to deal
+in the Software without restriction, including without limitation the rights
+to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+copies of the Software, and to permit persons to whom the Software is
+furnished to do so, subject to the following conditions:
+
+The above copyright notice and this permission notice shall be included in all
+copies or substantial portions of the Software.
+
+THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+SOFTWARE.
+*******************************************************************************/
+
+#ifndef CURRENT_SHERLOCK_REPLICATOR_H
+#define CURRENT_SHERLOCK_REPLICATOR_H
+
+#include <functional>
+#include <string>
+#include <thread>
+
+#include "exceptions.h"
+#include "sherlock.h"
+#include "stream_data.h"
+
+#include "../Blocks/HTTP/api.h"
+#include "../Blocks/SS/ss.h"
+
+#include "../Bricks/sync/scope_owned.h"
+#include "../Bricks/waitable_atomic/waitable_atomic.h"
+
+#include "../TypeSystem/Reflection/types.h"
+
+namespace current {
+namespace sherlock {
+
+template <typename STREAM_ENTRY>
+class SubscribableRemoteStream final {
+ public:
+  using entry_t = STREAM_ENTRY;
+
+  class RemoteStream final {
+   public:
+    RemoteStream(const std::string& url, const std::string& top_level_name, const std::string& namespace_name)
+        : url_(url),
+          schema_(Value<reflection::ReflectedTypeBase>(reflection::Reflector().ReflectType<entry_t>()).type_id,
+                  top_level_name,
+                  namespace_name) {}
+
+    void CheckSchema() const {
+      const auto response = HTTP(GET(url_ + "/schema.simple"));
+      if (static_cast<int>(response.code) == 200) {
```

Comment:
`static_cast<int>` is useful for unit tests to dump the "wrong" value as an integer. In real code, if (response.code == HTTPResponseCode.OK) is cleaner.